RollbarTests/PayloadTruncationTests.m

```diff
@@ -0,0 +1,137 @@
+
+
+    XCTAssertTrue(testStringBytes2 == (2 * testStringBytes1));
+    XCTAssertTrue(4 == testString1.length);
+    XCTAssertTrue(testString2.length == (2 * testString1.length));
+    XCTAssertTrue(testStringBytes1 == (4 * testString1.length));
+    XCTAssertTrue(testStringBytes2 == (4 * testString2.length));
+
+    XCTAssertTrue((4 * [RollbarPayloadTruncator measureTotalEncodingBytes:testString1
+                                                       usingEncoding:NSUTF8StringEncoding])
+                  == [RollbarPayloadTruncator measureTotalEncodingBytes:testString1
+                                                          usingEncoding:NSUTF32StringEncoding]);
+    XCTAssertTrue((4 * [RollbarPayloadTruncator measureTotalEncodingBytes:testString1])
+                  == [RollbarPayloadTruncator measureTotalEncodingBytes:testString1
+                                                          usingEncoding:NSUTF32StringEncoding]);
+}
+
+- (void)testTruncateStringToTotalBytes {
+    
+    NSString *testString = @"ABCDE-ABCDE-ABCDE";
+    const int truncationBytesLimit = 10;
+    XCTAssertTrue(truncationBytesLimit
+                  < [RollbarPayloadTruncator measureTotalEncodingBytes:testString]
+                  );
+    NSString *truncatedString = [RollbarPayloadTruncator truncateString:testString
+                                                           toTotalBytes:truncationBytesLimit];
+    XCTAssertTrue([RollbarPayloadTruncator measureTotalEncodingBytes:testString]
+                  > [RollbarPayloadTruncator measureTotalEncodingBytes:truncatedString]);
+    XCTAssertTrue(truncationBytesLimit
+                  >= [RollbarPayloadTruncator measureTotalEncodingBytes:truncatedString]
+                  );
+    XCTAssertTrue(testString.length > truncatedString.length);
+    
+    testString = @"abcd";
+    truncatedString = [RollbarPayloadTruncator truncateString:testString
+                                                 toTotalBytes:truncationBytesLimit];
+    XCTAssertTrue([RollbarPayloadTruncator measureTotalEncodingBytes:testString]
+                  == [RollbarPayloadTruncator measureTotalEncodingBytes:truncatedString]);
+    XCTAssertTrue(testString.length == truncatedString.length);
+    XCTAssertTrue(testString == truncatedString);
+}
+
+- (void)testPayloadTruncation {
+
+    @try {
+        NSArray *crew = [NSArray arrayWithObjects:
+                         @"Dave",
+                         @"Heywood",
+                         @"Frank", nil];
+        // This will throw an exception.
+        NSLog(@"%@", [crew objectAtIndex:10]);
+    }
+    @catch (NSException *exception) {
+        [Rollbar error:nil exception:exception];
+    }
+//    @catch (id exception) {
+//        [Rollbar error:@"GOT AN EXCEPTION" exception:exception];
+//    }
+    @finally {
+        NSLog(@"Cleaning up");
+    }
+    
+    [NSThread sleepForTimeInterval:1.0f];
+    NSArray *items = RollbarReadLogItemFromFile();
```

Comment:
This test is cheating by using the JSON serialization options to make everything mutable inside the items. At the one place we call `truncatePayload` the payload itself is a mutable dictionary because it is constructed that way but there is no guarantee that the values inside of it are mutable.

Reply:
yes, that was the only way (that i found) to get access to a real payload object to start testing the truncation strategies implementations. and, also, before i added itereateObjectStructure method.
actually, now i am thinking that even with all the recent changes to make all the containers within a payload mutable (while composing the payload), the client code can probably add some nonmutable once that we are not covering... should i go ahead and update itereateObjectStructure method implementation to account for those?

Reply:
Yeah I think that function needs to handle the possibility of getting arbitrary nested mutable and non-mutable items because the user can add things.

Reply:
done.